Task as struct, static non-async funcs, and UnsafeCurrentTask

ktoso opened the pull request:
- Applies the latest Task API revisions

----

based on https://forums.swift.org/t/pitch-2-structured-concurrency/43452/116 - detailed additional discussion here if necessary

ktoso committed:
Task as struct, static non-async funcs, and UnsafeCurrentTask

ktoso commented:
TODO: move back to "canceLLed"

ktoso commented:
Reverted the spelling change and this should be ready to go -- please have a look if looks right when you get the time @DougGregor

ktoso commented on proposals/nnnn-structured-concurrency.md:
````diff
@@ -377,13 +377,43 @@ Much of the proposed implementation of structured concurrency is in the APIs for
+
+```swift
+func foo() async { 
+  let task = await Task.current()
````
if Kavon's async properties would land this could be `await Task.current` -- more consistent with all the other APIs then

ktoso replied:
fyi @kavon

ktoso committed:
add to revision history

DougGregor commented on proposals/nnnn-structured-concurrency.md:
````diff
@@ -377,13 +377,43 @@ Much of the proposed implementation of structured concurrency is in the APIs for
+}
+```
+
+An asynchronous function _always_ executes within a task. As the `current` function itself is also asynchronous, it can only be invoked from an asynchronous function and is guaranteed to return the apropriate `Task`. The task object obtained this way is safe access by other tasks/threads. The APIs surfaced on it are specifically designed to also be safe, and relatively cheap, to be called from ot ...
````
```suggestion
An asynchronous function _always_ executes within a task. As the `current` function itself is also asynchronous, it can only be invoked from an asynchronous function and is guaranteed to return the appropriate `Task`. The task object obtained this way is safe access by other tasks/threads. The APIs surfaced on it are specifically designed to also be safe, and relatively cheap, to be called from other tasks.
```

DougGregor commented on proposals/nnnn-structured-concurrency.md:
```diff
@@ -377,13 +377,43 @@ Much of the proposed implementation of structured concurrency is in the APIs for
+
+An asynchronous function _always_ executes within a task. As the `current` function itself is also asynchronous, it can only be invoked from an asynchronous function and is guaranteed to return the apropriate `Task`. The task object obtained this way is safe access by other tasks/threads. The APIs surfaced on it are specifically designed to also be safe, and relatively cheap, to be called from ot ...
+
+Tasks are equatable and hashable, this can be used to store and compare them, and can be used e.g. to answer questions such as "is this the same task I was called from before" etc. Keep in mind though, that tasks should not be held onto unnecessarily.
```
```suggestion
Tasks are `Equatable` and `Hashable`, this can be used to store and compare them, and can be used e.g. to answer questions such as "is this the same task I was called from before" etc. Keep in mind though, that tasks should not be held onto unnecessarily.
```

DougGregor commented on proposals/nnnn-structured-concurrency.md:
````diff
@@ -411,11 +441,21 @@ extension Task {
 }
 ```
 
+The `currentPriority()` operation queries the priority of the currently-executing task. It may also be invoked from a synchronous context and returns `.default` if not running inside of a task in that case.
````
```suggestion
The `currentPriority` operation queries the priority of the currently-executing task. It may also be invoked from a synchronous context and returns `.default` if not running inside of a task in that case.
```

It's basically `Task.unsafeCurrent?.priority ?? .default`, then.

ktoso replied:
Right 👍

DougGregor commented on proposals/nnnn-structured-concurrency.md:
```diff
@@ -411,11 +441,21 @@ extension Task {
+
+Task priorities are set on task creation (e.g., `Task.runDetached` or `Task.Group.add`) and can be escalated later, e.g., if a higher-priority task waits on the task handle of a lower-priority task.
+
 The `currentPriority()` operation queries the priority of the currently-executing task. Task priorities are set on task creation (e.g., `Task.runDetached` or `Task.Group.add`) and can be escalated later, e.g., if a higher-priority task waits on the task handle of a lower-priority task.
```
Another `currentPriotity()` that needs to drop the parentheses.

ktoso replied:
Done thanks

DougGregor commented on proposals/nnnn-structured-concurrency.md:
````diff
@@ -377,13 +377,43 @@ Much of the proposed implementation of structured concurrency is in the APIs for
 ```swift
-@frozen enum Task { }
+func synchronous() {
+  if let task = Task.unsafeCurrent {
````
We're missing definitions of both `current` and `unsafeCurrent`.

DougGregor replied:
Also, can we get a `Task` from an `UnsafeCurrentTask`? (It seems like we should be able to). Can we get a `Task` from a `Task.Handle`? (It seems like we should be able to). In this case, we effectively have *three* copies of those `Task` APIs that are safe to use from effectively anywhere

ktoso replied:
> Also, can we get a Task from an UnsafeCurrentTask? (It seems like we should be able to). 

This seems reasonable; as Task has only "safe" APIs it makes sense to get it from an `UnsafeCurrentTask` 🤔 

Will amend this

> Can we get a Task from a Task.Handle? (It seems like we should be able to). 

I don't think that should be the case actually, but I can be convinced otherwise. It's not a core property and need of a "read side" of the work (the Task.Handle) to be able to query stuff about the `Task`. I know people "asked for that on the forums" but what's the actual reason? Why would a _read side_ ever query isCanceled or priority of ongoing work -- all it needs is to wait for the handle to complete, nothing else.

> In this case, we effectively have three copies of those Task APIs that are safe to use from effectively anywhere

DougGregor replied:
`Task.Handle` already has `isCancelled` and `cancel`. If you're holding a handle for some work, you might want to cancel it (if you no longer care about the result) or determine whether it's been cancelled (say, to drop that handle from the set of results you're still waiting for).

ktoso replied:
Okey, lemme add a `handle.task` then 👍

DougGregor commented on proposals/nnnn-structured-concurrency.md:
```diff
@@ -377,13 +377,43 @@ Much of the proposed implementation of structured concurrency is in the APIs for
+
+It is also possible to obtain a reference to the "current" task from _synchronous_ functions.
+
+This type is the core underlying type used to implement most of the APIs discussed in the following sections (which may be used from non-async contexts). Accessing using this API performs a thread-local lookup of a specific thread-local variable that is maintained by the Swift concurrency runtime.
```
This feels important enough that it should come as part of the Proposed Solution, not be buried here. The fact that synchronous code can check for cancellation when it's running on behalf of asynchronous code is important.

ktoso replied:
True, will add info into the proposed solution

ktoso replied:
Added a section

DougGregor commented on proposals/nnnn-structured-concurrency.md:
```diff
@@ -497,14 +537,18 @@ try await eat(dinnerHandle)
+  static var isCancelled: Bool
+  
+  /// Returns `true` if the task is cancelled, and should stop executing.
+  var isCancelled: Bool
```
I feel like we should split these two into different sections. The instance property is the "core" operation that comes up when describing the cancellation APIs, and the static property should be grouped with all of the other static operations on `Task` that can be queried from both synchronous and asynchronous code. The latter will have some rationale about the defaults chosen, for example, and why synchronous code should have access here.

ktoso replied:
Sounds good, will split it up

ktoso committed:
Apply suggestions from code review

Co-authored-by: Doug Gregor <dgregor@apple.com>

DougGregor commented:
I'm fine with going ahead and merging this and iterating on the main branch.

ktoso committed:
add separate section on synchronous access; define current... APIs

ktoso committed:
Merge branch 'wip-task-reinvented' of github.com:ktoso/swift-evolution into wip-task-reinvented

ktoso commented:
Thank you for review @DougGregor, I addressed all comments and added a new section on the specific access from synchronous code. 

The static functions got their own section and are expressed as sugar, which they are.

I also added `handle.task` and explained use cases.

Those changes are in here: https://github.com/DougGregor/swift-evolution/pull/48/commits/d57432660000f0068c305692a20f47a7bc4e83f5
